### PR TITLE
feat: add security_group to scaleway compute resource

### PR DESCRIPTION
**Pull request by abarbare**
##### SUMMARY

Let the scaleway user assigning a Security Group resource to a compute server

##### ISSUE TYPE

- Feature Pull Request

##### COMPONENT NAME

scaleway_compute
##### ANSIBLE VERSION

```paste below
ansible 2.8.0.dev0 (scw_compute_sg 8ff9e60012) last updated 2018/09/16 21:03:04 (GMT +200)
  config file = /Users/abarbare/.ansible.cfg
  configured module search path = [u'/Users/abarbare/.ansible/plugins/modules', u'/usr/share/ansible/plugins/modules']
  ansible python module location = /Users/abarbare/Documents/ansible/lib/ansible
  executable location = /Users/abarbare/Documents/ansible/bin/ansible
  python version = 2.7.15 (default, Jul 23 2018, 21:27:06) [GCC 4.2.1 Compatible Apple LLVM 9.1.0 (clang-902.0.39.2)]
```

### CODE REVIEW

**Comment by pilou-** [comment]
@abarbare a rebase is required here

**Comment by abarbare** [comment]
Cc @sieben

File: lib/ansible/modules/cloud/scaleway/scaleway_compute.py [hunk 1]
```diff
@@ -354,7 +376,7 @@ def present_strategy(compute_api, wished_server):
         if compute_api.module.check_mode:
             return changed, {"status": "Server %s attributes would be changed." % target_server["id"]}
 
-        server_change_attributes(compute_api=compute_api, target_server=target_server, wished_server=wished_server)
+        target_server = server_change_attributes(compute_api=compute_api, target_server=target_server, wished_server=wished_server)
```
Comment (pilou-) [inline lib/ansible/modules/cloud/scaleway/scaleway_compute.py:380]: I would rather use another variable name, maybe `created_server` ? `server_change_attributes` method should also use another variable name for the returned object.

File: lib/ansible/modules/cloud/scaleway/scaleway_compute.py [hunk 2]
```diff
@@ -127,6 +127,12 @@
     - Time to wait before every attempt to check the state of the server
     required: false
     default: 3
+
+  security_group:
```
Comment (pilou-) [inline lib/ansible/modules/cloud/scaleway/scaleway_compute.py:131]: If no value is provided, a default security group is automatically used, that might be mentioned here or in the `notes` section.

File: test/legacy/roles/scaleway_compute/tasks/security_group.yml [hunk 3]
```diff
@@ -0,0 +1,117 @@
...
+    wait: true
+
+  register: server_destroy_task
+
+- debug: var=server_destroy_task
+
+- assert:
+    that:
+      - server_destroy_task is success
+      - server_destroy_task is changed
```
Comment (remyleone) [inline test/legacy/roles/scaleway_compute/tasks/security_group.yml]: Could you add a new line?

File: lib/ansible/modules/cloud/scaleway/scaleway_compute.py [hunk 4]
```diff
@@ -575,29 +599,51 @@ def server_attributes_should_be_changed(compute_api, target_server, wished_serve
...
     compute_api.module.debug("Starting patching server attributes")
-    patch_payload = dict((x, wished_server[x])
-                         for x in PATCH_MUTABLE_SERVER_ATTRIBUTES
-                         if x in wished_server and x in target_server)
+    patch_payload = dict()
+
+    for key in PATCH_MUTABLE_SERVER_ATTRIBUTES:
+        if key in target_server and key in wished_server:
+            # When you are working with dict, only ID matter as we ask user to put only the resource ID in the playbook
+            if isinstance(target_server[key], dict) and "id" in target_server[key] and wished_server[key]:
```
Comment (remyleone) [inline lib/ansible/modules/cloud/scaleway/scaleway_compute.py:624]: I've add problem before with using isinstance. What if you have an AnsibleMapping there? Could issubclass be better?

File: lib/ansible/modules/cloud/scaleway/scaleway_compute.py [hunk 5]
```diff
@@ -271,30 +291,33 @@ def public_ip_payload(compute_api, public_ip):
...
+
+    response = compute_api.post(path="servers", data=data)
 
     if not response.ok:
         msg = 'Error during server creation: (%s) %s' % (response.status_code, response.json)
         compute_api.module.fail_json(msg=msg)
 
     try:
-        target_server = response.json["server"]
+        created_server = response.json["server"]
```
Comment (remyleone) [inline lib/ansible/modules/cloud/scaleway/scaleway_compute.py]: Is the refactoring of the name necessary? It adds a lot of lines in the review.

Comment (pilou-) [reply lib/ansible/modules/cloud/scaleway/scaleway_compute.py]: See my [previous review](https://github.com/ansible/ansible/pull/45699#pullrequestreview-157954221), I guess I wasn't clear enough: I asked only for:
```
created_server = server_change_attributes(compute_api=compute_api, target_server=target_server, wished_server=wished_server)
```

File: test/legacy/roles/scaleway_compute/tasks/security_group.yml (hunk 3 above)
Comment (abarbare) [reply test/legacy/roles/scaleway_compute/tasks/security_group.yml]: Well catch ! Thanks 👍

File: lib/ansible/modules/cloud/scaleway/scaleway_compute.py (hunk 1 above)
Comment (abarbare) [reply lib/ansible/modules/cloud/scaleway/scaleway_compute.py:380]: Actually, I can't modify only this variable as it is returned by the function. 
The function will return `target_server` for existing server or updated one. Changing only this variable will cause an undefined variable error on return.

File: lib/ansible/modules/cloud/scaleway/scaleway_compute.py (hunk 5 above)
Comment (abarbare) [reply lib/ansible/modules/cloud/scaleway/scaleway_compute.py]: I added the [explantation](https://github.com/ansible/ansible/pull/45699#discussion_r221433211)

File: lib/ansible/modules/cloud/scaleway/scaleway_compute.py (hunk 4 above)
Comment (abarbare) [reply lib/ansible/modules/cloud/scaleway/scaleway_compute.py:624]: I tried to use issubclass but as I compare result of the API, I have to cast my variable using `type` which is not allowed by the CI.
The data will always be known data as we get them through the API.

**Comment by remyleone** [comment]
LGTM

**Comment by abarbare** [comment]
Rebased to relauch CI.

**Comment by pilou-** [comment]
@abarbare another rebase is needed

**Comment by abarbare** [comment]
Rebased to relauch CI.

**Comment by abarbare** [comment]
shipit

**Comment by remyleone** [comment]
LGTM

**Comment by abarbare** [comment]
bot_status

**Pull request merged**